rules/Php83/Rector/ClassConst/AddTypeToConstRector.php

```diff
@@ -20,19 +20,29 @@
 use PhpParser\Node\Stmt\ClassConst;
 use PHPStan\Reflection\ClassReflection;
 use PHPStan\Reflection\ReflectionProvider;
+use Rector\Contract\Rector\ConfigurableRectorInterface;
 use Rector\PHPStanStaticTypeMapper\Enum\TypeKind;
 use Rector\Rector\AbstractRector;
 use Rector\StaticTypeMapper\StaticTypeMapper;
 use Rector\ValueObject\PhpVersionFeature;
 use Rector\VersionBonding\Contract\MinPhpVersionInterface;
-use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\ConfiguredCodeSample;
 use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+use Webmozart\Assert\Assert;
 
 /**
  * @see \Rector\Tests\Php83\Rector\ClassConst\AddTypeToConstRector\AddTypeToConstRectorTest
  */
-final class AddTypeToConstRector extends AbstractRector implements MinPhpVersionInterface
+final class AddTypeToConstRector extends AbstractRector implements ConfigurableRectorInterface, MinPhpVersionInterface
 {
+    /**
+     * @api
+     * @var string
+     */
+    final public const IGNORE_INHERITANCE = 'ignore_inheritance';
```

Comment:
I prefer more verbose config, like `allow_has_child`, then, add check `FamilyRelationsAnalyzer->getChildrenOfClassReflection()` that collect children on the `paths()` provided in `rector.php` config, and check if it has constant:

```php
// resolve current class Reflection ....
$classReflection = $this->reflectionResolver->resolveClassReflection($node);
if (! $classReflection instanceof \PHPStan\Reflection\ClassReflection)
    return null;
}

// verify has children with constant
if ($this->allow_has_child) {
    $childClasses = $this->familyRelationsAnalyzer->getChildrenOfClassReflection($classReflection);
    $constName = $this->getName($classConst->name);
    foreach ($childClasses as $childClass) {
        if ($childClass->hasConstant($constName)) {
             continue 2; // don't change, continue to inner loop...
        }
    }
}
```

Above is pseudo code, but I think you have the idea ;)

Reply:
Thanks for the quick feedback @samsonasik!

I'm not entirely sure what you're getting at. Do you want it to only type the constant if a child class does not define the same constant?

Reply:
Yes, that means constant overridden, which will make fatal error. After some thinking:

```php
$childClass->hasConstant($constName);
```

seems will be always `true` on `public` or `protected` constant, so I think you need to get real `Class_` via `AstResolver`

```php
foreach ($childClasses as $childClass) {
        $classLike = $this->astResolver->resolveClassFromClassReflection($childClass);
        if ($childClassLike instanceof \PhpParser\Node\Stmt\ClassLike) {
            continue;
        }
        
        foreach ($childClassLike->getConstants() as $constant) {
             foreach ($constant->const as $childConst) {
                    if ($this->nodeComparator->areNodesEquals($childConst, $constNode)) {
                          continue 3; // don't change, continue to inner loop...
                    }
             } 
        }
    }
```

When `allow_has_child` enabled:

- If child has constant, **but not exists**, it should be allowed when constant names are different, see https://3v4l.org/HcqRB
- if child has constant, **and equal**, it should not be allowed, even value type is equal, see
        -  https://3v4l.org/SUXMa
        - https://3v4l.org/rFOhl

For note: the `allow_has_child` should be configurable with default `false`.

---

On personal note, I prefer the original behaviour for this.

Reply:
Another way that might work is by use its `ClassConstantReflection` object itself, you can add fixture test with both parent and child in same fixture file, then you can check something like this:

```php
foreach ($childClasses as $childClass) {
        if (! $childClass->hasConstant($constName)) {
            continue;
        }
          
        $constantChild =   $childClass->getConstant($constName);
        // debug here, check whether it own by child class or only deriving...
        if ($constantChild->getFileName() !== $classReflection->getFileName()) {
            continue 2; // don't change, continue to inner loop...
      } 
    }
```

Reply:
Thanks for the help @samsonasik, I've implemented these changes now.

It will now _only_ add the type if:
- there are no child classes
- the child classes _do not_ override the constant

Reply:
Thank you @ndench, I will check when I have a chance, and verify the possible improvement if needed.

This may not be our priority for now, so this may be here for a while for some thinking and/or discussion with @TomasVotruba in our next meeting, so you can create your custom rule for your needs for now.